src/ipnet.rs

```diff
@@ -0,0 +1,720 @@
+    pub fn contains(&self, addr: IpAddr) -> bool {
+        match (self, addr) {
+            (IpNet::V4(net), IpAddr::V4(ip)) => net.contains(ip),
+            (IpNet::V6(net), IpAddr::V6(ip)) => net.contains(ip),
+            (_, _) => false,
+        }
+    }
+}
+
+impl From<Ipv4Net> for IpNet {
+    fn from(n: Ipv4Net) -> IpNet {
+        IpNet::V4(n)
+    }
+}
+
+impl From<Ipv6Net> for IpNet {
+    fn from(n: Ipv6Net) -> IpNet {
+        IpNet::V6(n)
+    }
+}
+
+impl std::fmt::Display for IpNet {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            IpNet::V4(inner) => write!(f, "{}", inner),
+            IpNet::V6(inner) => write!(f, "{}", inner),
+        }
+    }
+}
+
+impl std::str::FromStr for IpNet {
+    type Err = IpNetParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let Some(ii) = s.find('/') else {
+            return Err(IpNetParseError::MissingSlash);
+        };
+
+        let addr_str = &s[..ii];
+        let prefix_str = &s[ii + 1..];
+
+        let prefix = prefix_str.parse().map_err(IpNetParseError::InvalidPrefix)?;
+        let addr = addr_str.parse().map_err(IpNetParseError::InvalidAddr)?;
+        IpNet::new(addr, prefix).map_err(IpNetParseError::PrefixValue)
+    }
+}
+
+#[cfg(feature = "schemars")]
+impl schemars::JsonSchema for IpNet {
+    fn schema_name() -> String {
+        "IpNet".to_string()
+    }
+
+    fn json_schema(gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
+        use crate::schema_util::label_schema;
+        schemars::schema::SchemaObject {
+            subschemas: Some(Box::new(schemars::schema::SubschemaValidation {
+                one_of: Some(vec![
+                    label_schema("v4", gen.subschema_for::<Ipv4Net>()),
+                    label_schema("v6", gen.subschema_for::<Ipv6Net>()),
+                ]),
+                ..Default::default()
+            })),
+            extensions: crate::schema_util::extension("IpNet", "0.1.0"),
+            ..Default::default()
+        }
+        .into()
+    }
+}
+
+pub const IPV4_NET_PREFIX_MAX: u8 = 32;
+
+/// An IPv4 subnet
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Ipv4Net {
+    addr: Ipv4Addr,
+    prefix: u8,
+}
+
+impl Ipv4Net {
+    pub fn new(addr: Ipv4Addr, prefix: u8) -> Result<Self, IpNetPrefixError> {
+        if prefix > IPV4_NET_PREFIX_MAX {
+            Err(IpNetPrefixError(prefix))
+        } else {
+            Ok(Self { addr, prefix })
+        }
+    }
+    pub fn host_net(addr: Ipv4Addr) -> Self {
+        Self {
+            addr,
+            prefix: IPV4_NET_PREFIX_MAX,
+        }
+    }
+    pub fn is_host_net(&self) -> bool {
+        self.prefix == IPV4_NET_PREFIX_MAX
+    }
+    pub fn addr(&self) -> Ipv4Addr {
+        self.addr
+    }
+
+    pub fn prefix(&self) -> u8 {
+        self.prefix
+    }
+
+    pub fn size(&self) -> u32 {
+        1u32 << self.prefix
```

Comment:
This should be `1u32 << (IPV4_NET_PREFIX_MAX - self.prefix)`

Reply:
egad; thanks